### ISSUE TITLE
Lua style custom allocators

### ISSUE THREAD

**Issue by awsdert** [HappySeaFox/sail#77]
Tried a search for alloc but that brought up a lot of junk results that didn't match what I was looking for, when lua is first initialised in C it provides the option of setting a custom allocator and user data where the allocator is of the form `void * Alloc( void *ud, void *ptr, size_t size, size_t want )` where ud is ignored by the default allocator and only used by a user supplied allocator, this is helpful when multi-threading as I can pass the threads object to the ud parameter which can then communicate the desire for memory to the main thread which handles the allocations instead, avoiding potential collisions when 2 or more threads try to re-allocate memory as they can just all have the same pointer & size reported to them after each re-allocation request is processed, also makes garbage collection easier.

Does SAIL have something similar? If not I'll have to write my own custom image loading functions, I'd prefer to not do that though as I'm trying to keep my graphic studies to just the graphics and not the formats (only recently learned how to use index buffers and am now trying to learn how to use textures).

**Comment by HappySeaFox** [comment]
Hi!

TL;DR Unfortunately, SAIL doesn't provide this feature currently.

Long read:

There are two types of allocations in SAIL:

1. Allocations made by SAIL in `libsail`, codecs, and other our libs. They're all made by our own allocation functions which are eventually just wrappers around malloc/free: https://github.com/HappySeaFox/sail/blob/master/src/libsail-common/utils.c#L887 . Unfortunately, there is no mechanism to override them currently w/o patching and distributing your own copy of SAIL.

2. Allocations made by underlying codecs like `libjpeg`. Some codecs like `JPEG` still use third-party libs like `libjpeg`, `libpng` etc. We don't control these allocations at all.

**Comment by awsdert** [comment]
Well that's disappointing, guess I'll just have to write my own then :|

**Comment by HappySeaFox** [comment]
It would be great if you share details on how exactly will you implement it! For example, libjpeg doesn't provide this feature as well.

**Comment by awsdert** [comment]
Well for now I gave up on jpeg as I couldn't find information on how big it's various details are (byte wise I mean), instead I'm going with PNG for now which was easy enough to read out all chunks (haven't gotten to the point of understanding each chunk yet), here's what the header looks like:

```
#ifndef IMG_PNG_H
#define IMG_PNG_H

#include "common.h"

typedef struct _IMG_PNG_ENTRY
{
	uint CRC;
	char type[5];
	BUFF data;
} IMG_PNG_ENTRY;

typedef struct _IMG_PNG_ENTRY_IHDR
{
	uint	SpanX;
	uint	SpanY;
	uchar	Depth;
	uchar	Color;
	uchar	CompressionMethod;
	uchar	FilterMethod;
	uchar	InterlaceMethod;
} IMG_PNG_ENTRY_IHDR;

typedef struct _IMG_PNG
{
	BIN_FILE	BinFile;
	BUFF		Entries;
} IMG_PNG;

IMG_PNG * alloc_img_png( ALLOC *Alloc, IMG_PNG *Image, tch const * const path );
IMG_PNG * load_img_png( ALLOC *Alloc, IMG_PNG *Image );
void empty_img_png( ALLOC *Alloc, IMG_PNG *Image );

#endif
```
The point of alloc/empty_img_png is to speed up the loading of textures on the fly, instead of destroying the PNG object and it's buffers when letting go of a texture it instead allows you to just outright change the file path stored and load a new texture for it, the reason the file
path is done in another function is to simplify the design of the underlying BIN_FILE object that's being relied on (which also uses the ALLOC object). The ALLOC object is just to simplify passing around both the `void *ud;` and the `void * (*Alloc)( void *, void *, size_t, size_t );` inside it, yes I used the Lua style allocator callback as the inspiration, didn't want an entire context for it like lua's `lua_State* L`, that has it's purpose but for this code that's just over the top so ALLOC was born. I'll be uploading all my code to gitlab.com/awsdert/glengine and eventually convert it for use in gitlab.com/awsdert/paw (which then be used to cleanup glengine), I'm designing the code so that all functions require the object to already exist, so whether it is part of another object or allocated on the fly it must exist prior to the calls, the reason some return the object is for convenience, for example:
`BIN_FILE * BinFile = alloc_bin_file( Alloc, &(Image->BinFile), path );` would give a convenient way to simultaneously check for success and get a shorter name to use in the function it is declared in. Btw I haven't uploaded the above code yet, I'm still looking through the format details atm.

**Comment by awsdert** [comment]
As someone who has experience in reading/writing PNG chunks, particularly EXIF metadata (I feel like my current method is wrong here), can you think of any issues that would crop up from this:
```
#ifndef IMG_PNG_H
#define IMG_PNG_H

#include "common.h"

typedef struct _IMG_PNG_ENTRY
{
	uint CRC;
	char type[5];
	BUFF data;
} IMG_PNG_ENTRY;

#define IMG_PNG_CLUT_GREYSCALE 0
#define IMG_PNG_CLUT_TRUECOLOR 2
#define IMG_PNG_CLUT_PALLETEPOS 3
#define IMG_PNG_CLUT_GREYSCALEA 4
#define IMG_PNG_CLUT_TRUECOLORA 6

typedef enum _IMG_PNG_CHUNK_TYPE
{
	IMG_PNG_CHUNK_bKGD,
	IMG_PNG_CHUNK_cHRM,
	IMG_PNG_CHUNK_dSIG,
	IMG_PNG_CHUNK_eXIf,
	IMG_PNG_CHUNK_iCCP,
	IMG_PNG_CHUNK_iTXt,
	IMG_PNG_CHUNK_pHYs,
	IMG_PNG_CHUNK_sBIT,
	IMG_PNG_CHUNK_sPLT,
	IMG_PNG_CHUNK_sRGB,
	IMG_PNG_CHUNK_sTER,
	IMG_PNG_CHUNK_tEXt,
	IMG_PNG_CHUNK_tRNS,
	IMG_PNG_CHUNK_zTXt,
	IMG_PNG_CHUNK_COUNT
} IMG_PNG_CHUNK_TYPE;

const char const * img_png_chunk_desc( IMG_PNG_CHUNK_TYPE type );

typedef struct _IMG_PNG
{
	uint		SpanX;
	uint		SpanY;
	uchar		Depth;
	uchar		Color;
	uchar		CompressionMethod;
	uchar		FilterMethod;
	uchar		InterlaceMethod;
	time_t		TimeLastChanged;
	float		Gama;
	BIN_FILE	BinFile;
	IMG_PNG_ENTRY Entry;
	BUFF		Palette;
	BUFF		AlternativePalette;
	BUFF		Chunks[IMG_PNG_CHUNK_COUNT];
	uint		Pixels;
	BUFF		FullImg;
} IMG_PNG;

IMG_PNG *
	setup_img_png( ALLOC *Alloc, IMG_PNG *Image, tch const * const path );
IMG_PNG *
	load_img_png( ALLOC *Alloc, IMG_PNG *Image, bool loopUntilRdAll );
void
	empty_img_png( ALLOC *Alloc, IMG_PNG *Image );

#endif
```

**Comment by HappySeaFox** [comment]
Re-opening temporarily for a possible discussion.

**Comment by awsdert** [comment]
Managed to get the basics down for PNG loading, just struggling with the scanline stuff, until I get that first palette based PNG loaded and printed to a PPM correctly I can't continue to advance my texture studies/experiments. Anyone who want's to help out I started a thread here:
https://cboard.cprogramming.com/c-programming/180440-way-do-scan-lines.html

Don't wanna side track this topic with PNG stuff after all

**Comment by HappySeaFox** [comment]
Closing for inactivity
